runtime/sema/check_member_expression.go

```diff
@@ -83,30 +83,78 @@ func (checker *Checker) VisitMemberExpression(expression *ast.MemberExpression)
 
 	// If the member access is optional chaining, only wrap the result value
 	// in an optional, if it is not already an optional value
-
 	if isOptional {
 		if _, ok := memberType.(*OptionalType); !ok {
-			return &OptionalType{Type: memberType}
+			memberType = &OptionalType{Type: memberType}
 		}
 	}
 
 	return memberType
 }
 
+// getReferenceType Returns a reference type to a given type.
+// Reference to an optional should return an optional reference.
+// This has to be done recursively for nested optionals.
+// e.g.1: Given type T, this method returns &T.
+// e.g.2: Given T?, this returns (&T)?
+func (checker *Checker) getReferenceType(typ Type) Type {
+	if optionalType, ok := typ.(*OptionalType); ok {
+		return &OptionalType{
+			Type: checker.getReferenceType(optionalType.Type),
+		}
+	}
+
+	return NewReferenceType(checker.memoryGauge, typ, UnauthorizedAccess)
+}
+
+func shouldReturnReference(parentType, memberType Type) bool {
+	if !isReferenceType(parentType) {
+		return false
+	}
+
+	return isContainerType(memberType)
+}
+
+func isReferenceType(typ Type) bool {
+	unwrappedType := UnwrapOptionalType(typ)
+	_, isReference := unwrappedType.(*ReferenceType)
+	return isReference
+}
+
+func isContainerType(typ Type) bool {
+	switch typ := typ.(type) {
+	case *CompositeType,
```

Comment:
Shouldn't we return references when the member type is an interface or restricted type as well?

Reply:
That's a good flag! Realized we would need to do it for all built-in types as well. Moved this logic to the type definitions itself: 9a64a7fddffd31976ee9abc25caaaf650083fd68